recipes-kernel/linux/4.19/patches/thorough-reset-interface-to-pciback-s-sysfs.patch

```diff
@@ -0,0 +1,542 @@
++
++		//If the given device is owned by pciback...
++		if (psdev->dev == dev) {
++
++			//mark it as a pciback device.
++			device_owned_by_pciback = true;
++
++			//If we have a physical device associated with the pciback device,
++			//mark this device as in-use.
++			if (psdev->pdev)
++				arg->use_count++;
++
++			//Stop searching; we've found a the PCIback device associated with this one.
++			break;
++		}
++	}
++
++	//Release the PCI device lock...
++	spin_unlock_irqrestore(&pcistub_devices_lock, flags);
++
++	//... and report if we've found a device that's not owned by pciback.
++	dev_dbg(&dev->dev, "%s\n", device_owned_by_pciback ? "is owned by pciback, and can be reset if not in use."
++			: "not owned by pciback, and thus cannot be reset.");
++
++	//If we've found a device that's not owned by pciback, update our data
++	//argument so it points to the most recent unowned device. (We check
++	//this like a flag, later: if it's never set, no one owns the device!)
++	if (!device_owned_by_pciback)
++		arg->last_non_pciback_device = dev;
++
++	//If we've found a device that's not owned by pciback, return false--
++	//this indicates that pci_walk_bus should cease its walk.
++	return !device_owned_by_pciback;
++}
++
++
++/**
++ * Returns true iff it should be safe to issue a secondary bus reset
++ * to the device; that is, if an SBR can be issued without disrupting
++ * other devices.
++ */
++static bool safe_to_sbr_device(struct pci_dev *dev)
++{
++	struct safe_to_sbr_arguments walk_result = { .last_non_pciback_device = NULL, .use_count = 0 };
++
++	//Walk the PCI bus, attempting to find if any of the given devices
++	pci_walk_bus(dev->bus, __safe_to_sbr_device_callback, &walk_result);
++
++	//If the device is in use, emit a warning error.
++	if(walk_result.use_count > 0)
++		dev_dbg(&dev->dev, "is in use; currently not safe to SBR device.\n");
++
++	//Return true iff we did not pick up any other devices
++	//that were either in use, or not owned by pciback.
++	return (walk_result.last_non_pciback_device == NULL) && (walk_result.use_count == 0);
++}
++
++
++/**
++ * Attempt a raw reset of the provided PCI device-- via any
++ * method available to us. This method prefers the gentlest
++ * possible reset method-- currently an FLR, which many
++ * PCIe devices should support.
++ *
++ * @param dev The pci device to be reset.
++ * @return Zero on success, or the error code generated by the reset method on failure.
++ */
++static int __pcistub_raw_device_reset(struct pci_dev *dev)
++{
++	//Determine if bus resetting techniques (SBR, slot resets)
++	//are safe, and thus should be allowed.
++	int allow_bus_reset = safe_to_sbr_device(dev);
++
++	//If FLRs are supported; we'll try to let the linux kernel
++	//manually reset the device.
++	if(device_supports_flr(dev)) {
++		dev_dbg(&dev->dev, "Resetting device using an FLR.");
++		return pci_reset_function(dev);
++	}
++
++	//Next, we'll try the next gentlest: a hotplugging reset
```

Comment:
This "gentlest" solution is the exact same as the "most drastic" one below.
`pci_reset_bus()` seems to now try a slot reset first, so I get where you coming from, but those 2 `if`s should probably be merged and the comments and debug messages fixed up...

Reply:
With `pci_reset_bus()` determines if slot reset is possible using `pci_probe_reset_slot()`, which is what `device_supports_slot_reset()`, introduced by this patch does. It seems this can just become:
```c
if (allow_bus_reset)
    return pci_reset_bus(dev);
```
... With the helpers `device_supports_slot_reset()` and `device_supports_bus_reset()` no longer being necessary.

Reply:
update coming soon. +1